### PR TITLE
Engine - Processing logic

### PR DESCRIPTION
**Pull request by JcabreraC**
|Related issue|
|---|
|#12633|

We need to align the semantics defined in our design document with the software we've created. Also, this software must be able to fit in the place of analysisd. Because of that, we've identified some friction points which we must address in order to operate within the rest of the Wazuh platform.

### Processing logic

Each graph node and each set of operations should behave as stated in the functional document. We need to implement some
logical combinations at the graph and operation levels. At least we need an exclusive or (XOR) ideally with a sorted list by the number of times an option has been selected.

We can implement this without own code, instead of using RXCPP primitives or combinators.

### CODE REVIEW

**Comment by Hamza1-coder** [comment]
I am new here any issue for me ?

**Review by juliancnn — APPROVED** [review]
LGTM :rocket:

**Comment by juliancnn** [comment]
## Acceptance test

With the acceptance tests, can be seen that there is no significant impact of EPS. Engine continues to respond in the same way.

|EPS | Debug | Release
|----|----|----|
|Master|788201|3937131|
|PR|739544|3912738|

image


[assets_db](https://github.com/wazuh/wazuh/files/8590845/assets_db.zip)
[RAW Results](https://github.com/wazuh/wazuh/files/8590878/RAW.Results.txt)

File: src/engine/source/builder/builderTypes.hpp [hunk 1]
```diff
@@ -17,26 +17,21 @@
 
 #include <rxcpp/rx.hpp>
 
+//#include <eventHandler.hpp>
+#include <baseTypes.hpp>
 #include "connectable.hpp"
```
Comment (nmkoremblum) [inline src/engine/source/builder/builderTypes.hpp:22]:
```suggestion
#include "baseTypes.hpp"
#include "connectable.hpp"

#include <rxcpp/rx.hpp>
```

**Review by nmkoremblum — CHANGES_REQUESTED** [review]

File: src/engine/source/base/baseTypes.hpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+#include "eventHandler.hpp"
+
+namespace base
+{
+    using Event = std::shared_ptr<EventHandler>;
+    using Document = json::Document;
+    using DocumentValue = json::Value; // TODO Se usa??
+    using Observable = rxcpp::observable<Event>;
+    using Lifter = std::function<Observable(Observable)>;
```
[the remark this answers is not in the document]
Comment (juliancnn) [reply src/engine/source/base/baseTypes.hpp:14]: We will not use `using json::Document` in a header file

File: src/engine/source/base/eventHandler.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#ifndef _H_EVENT_HANDLER
+#define _H_EVENT_HANDLER
+
+#include <json.hpp>
```
[the remark this answers is not in the document]
Comment (juliancnn) [reply src/engine/source/base/eventHandler.hpp:4]: for internal libraries we use <>

File: src/engine/source/base/eventHandler.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace base
+{
+
+class EventHandler
+{
+
+private:
+    // Control
+    bool is_decoded; ///< True if it reached the end of the decoding stage
```
[the remark this answers is not in the document]
Comment (juliancnn) [reply src/engine/source/base/eventHandler.hpp:14]: :+1:

File: src/engine/source/builder/builderTypes.hpp (hunk 1 above)
Comment (juliancnn) [reply src/engine/source/builder/builderTypes.hpp:22]: Should be:
```cpp
#include <rxcpp/rx.hpp>

#include <baseTypes.hpp>

#include "connectable.hpp"
```

**Review by nmkoremblum — APPROVED** [review]
LGTM :)

**Pull request merged**